src/oauth2.h

```diff
@@ -31,13 +31,42 @@ constexpr const char kDefaultTokenEndpoint[] =
   "https://www.googleapis.com/oauth2/v3/token";
 }
 
+class Expiration {
+ public:
+  virtual bool IsExpired(std::chrono::seconds slack) = 0;
+  virtual void Reset(std::chrono::seconds duration) = 0;
+};
+
+template<typename Clock>
+class ExpirationImpl : public Expiration {
+ public:
+  bool IsExpired(std::chrono::seconds slack) override {
+    return token_expiration_ < Clock::now() + slack;
+  }
+
+  void Reset(std::chrono::seconds duration) override {
+    token_expiration_ = Clock::now() + duration;
+  }
+
+ private:
+  typename Clock::time_point token_expiration_;
+};
+
 class OAuth2 {
  public:
   OAuth2(const Environment& environment)
-    : environment_(environment), token_endpoint_(kDefaultTokenEndpoint) {}
+    : OAuth2(environment,
+             std::unique_ptr<Expiration>(
+                 new ExpirationImpl<std::chrono::system_clock>())) {}
 
   std::string GetAuthHeaderValue();
 
+ protected:
+  OAuth2(const Environment& environment, std::unique_ptr<Expiration> expiration)
```

Comment:
[Optional] Have you considered adding a `SetExpirationForTest` method instead? Was there a reason to prefer this approach over that one?